plugins/antispam.py

```diff
@@ -0,0 +1,553 @@
+                        except Forbidden:
+                            self.parent.logger.warning(f"Can't edit member {self.member.display_name}")
+                        else:
+                            self.muted = False
+                        break
+            if time.time() - self.update_time > t_config['infraction_timeout']:
+                self.infractions = max(0, self.infractions-1)
+                self.update_time = time.time()
+
+    async def activate(self):
+        self.run_timer = True
+        self.s_thresholds = {}
+        if "members" not in self.storage:
+            self.storage["members"] = {}
+        for guild in self.client.guilds:
+            if guild.id not in self.storage["members"]:
+                self.storage["members"][guild.id] = {}
+            else:
+                for k, t_member in self.storage["members"][guild.id].items():
+                    t_member.reinit(self)
+            if str(guild.id) not in self.plugin_config:
+                self.plugin_config[str(guild.id)] = self.plugin_config["default"]
+            self.calc_thresholds(guild)
+
+        loop = asyncio.new_event_loop()
+        t_loop = asyncio.get_event_loop()
+        self.timer = threading.Thread(target=self.start_timer, args=[loop, t_loop])
+        self.timer.setDaemon(True)
+        self.timer.start()
+
+    async def deactivate(self):
+        self.run_timer = False
+
+    # Event handlers
+
+    async def on_guild_join(self, guild):
+        self.storage["muted_members"][str(guild.id)] = {}
+        self.plugin_config[str(guild.id)] = self.plugin_config["default"]
+
+    async def on_guild_remove(self, guild):
+        self.storage["muted_members"].pop(str(guild.id))
+        self.plugin_config.pop(str(guild.id))
+
+    async def on_message(self, msg):
+        """
+        Counts the messages in the time period.
+        First element is time.time(), second is the count
+        """
+        if msg.author == msg.guild.me:
+            return
+
+        if msg.author.id in self.storage["members"][msg.guild.id]:
+            t_member = self.storage["members"][msg.guild.id][msg.author.id]
+            t_config = self.plugin_config[str(msg.guild.id)]
+            if time.time() - t_member.update_time < t_config["timeout"]:
+                t_member.messages += 1
+                if t_member.messages > t_config["message_count"]:
+                    if t_member.messages % t_config["message_count"] == 0:
+                        await t_member.infract()
+                    else:
+                        t_member.update_time = time.time()
+                    if t_member.infractions >= self.s_thresholds[msg.guild.id][1] and t_config["spam_delete"]:
+                        await msg.delete(reason="Spam filtering.")
+                    elif t_member.infractions >= self.s_thresholds[msg.guild.id][0] and t_config["spam_reaction"]:
+                        if len(t_config["spam_reaction"]) != 18:
```

Comment:
Most `snowflake`s are 18 digits, but they can be more (up to 20 digits) or less, so it's best we don't hardcode 18.